README.md
axslogparser
=======

[![Build Status](https://travis-ci.org/Songmu/axslogparser.png?branch=master)][travis]
[![Coverage Status](https://coveralls.io/repos/Songmu/axslogparser/badge.png?branch=master)][coveralls]
[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)][license]
[![GoDoc](https://godoc.org/github.com/Songmu/axslogparser?status.svg)][godoc]

[travis]: https://travis-ci.org/Songmu/axslogparser
[coveralls]: https://coveralls.io/r/Songmu/axslogparser?branch=master
[license]: https://github.com/Songmu/axslogparser/blob/master/LICENSE
[godoc]: https://godoc.org/github.com/Songmu/axslogparser

## Description

An accesslog parser supports apache log (common and combined) and ltsv accesslog (http://ltsv.org).

## Supported Formats

- Apache Logs (also cared in the case of using tab character as delimiter)
  - Common log format
  - Common log format with vhost
  - Combined log format
  - Combined log format with extra fields
- LTSV Accesslog format according to http://ltsv.org

## Author

[Songmu](https://github.com/Songmu)


CHANGELOG.md
# Changelog

## [v1.4.0](https://github.com/Songmu/axslogparser/compare/v1.3.0...v1.4.0) (2020-06-24)

* Make parsers used in GuessParser and Parse configurable [#22](https://github.com/Songmu/axslogparser/pull/22) ([susisu](https://github.com/susisu))
* update travis settings [#21](https://github.com/Songmu/axslogparser/pull/21) ([Songmu](https://github.com/Songmu))

## [v1.3.0](https://github.com/Songmu/axslogparser/compare/v1.2.0...v1.3.0) (2020-04-17)

* Ignore LTSV unmarshal error if loose [#20](https://github.com/Songmu/axslogparser/pull/20) ([susisu](https://github.com/susisu))

## [v1.2.0](https://github.com/Songmu/axslogparser/compare/v1.1.0...v1.2.0) (2019-06-03)

* Add Loose option to ignore non-fatal formatting errors [#18](https://github.com/Songmu/axslogparser/pull/18) ([motemen](https://github.com/motemen))
* remove unused %s in errors.Wrap [#16](https://github.com/Songmu/axslogparser/pull/16) ([itchyny](https://github.com/itchyny))

## [v1.1.0](https://github.com/Songmu/axslogparser/compare/v1.0.0...v1.1.0) (2018-08-29)

* [incompatible] [bugfix] s/RemoteUser/RemoteLogname/g [#14](https://github.com/Songmu/axslogparser/pull/14) ([Songmu](https://github.com/Songmu))
* care remote_user which contains whitespace in apache log [#13](https://github.com/Songmu/axslogparser/pull/13) ([Songmu](https://github.com/Songmu))
* Support to multi ips into a remote host field [#12](https://github.com/Songmu/axslogparser/pull/12) ([softctrl](https://github.com/softctrl))

## [v1.0.0](https://github.com/Songmu/axslogparser/compare/5cfe5b4ad944...v1.0.0) (2018-04-04)

* Accept other than '-' as RemoteUser [#11](https://github.com/Songmu/axslogparser/pull/11) ([ulyssessouza](https://github.com/ulyssessouza))
* Fix typo in error message [#9](https://github.com/Songmu/axslogparser/pull/9) ([yano3](https://github.com/yano3))
* Consider the case where a host field is at the beginning of the LTSV log in `GuessParser` [#8](https://github.com/Songmu/axslogparser/pull/8) ([Songmu](https://github.com/Songmu))
* Readme [#7](https://github.com/Songmu/axslogparser/pull/7) ([Songmu](https://github.com/Songmu))
* add tests for LTSV [#6](https://github.com/Songmu/axslogparser/pull/6) ([Songmu](https://github.com/Songmu))
* enhance testing [#5](https://github.com/Songmu/axslogparser/pull/5) ([Songmu](https://github.com/Songmu))
* better error handling [#4](https://github.com/Songmu/axslogparser/pull/4) ([Songmu](https://github.com/Songmu))
* Adjust field names [#3](https://github.com/Songmu/axslogparser/pull/3) ([Songmu](https://github.com/Songmu))
* Adjust LTSV parser [#2](https://github.com/Songmu/axslogparser/pull/2) ([Songmu](https://github.com/Songmu))
* Better parsing for apache log [#1](https://github.com/Songmu/axslogparser/pull/1) ([Songmu](https://github.com/Songmu))


ltsv_test.go
package axslogparser

import (
	"strings"
	"testing"
)

var parseLTSVErrorTests = []struct {
	Name           string
	Input          string
	ContainsString string
	HardError      bool
}{
	{
		Name:           "broken",
		Input:          "hoge",
		ContainsString: "(not a ltsv)",
		HardError:      true,
	},
	{
		Name: "invalid request",
		Input: "time:08/Mar/2017:14:12:40 +0900\t" +
			"host:192.0.2.1\t" +
			"req:POST /api/v0/tsdb hoge HTTP/1.1\t" +
			"status:200\t" +
			"size:36\t" +
			"ua:mackerel-agent/0.31.2 (Revision 775fad2)\t" +
			"reqtime:0.087\t" +
			"taken_sec:0.087\t" +
			"vhost:mackerel.io",
		ContainsString: "(invalid request)",
	},
	{
		Name: "unmarshal error",
		Input: "time:08/Mar/2017:14:12:40 +0900\t" +
			"host:192.0.2.1\t" +
			"req:POST /api/v0/tsdb HTTP/1.1\t" +
			"status:200\t" +
			"size:36\t" +
			"ua:mackerel-agent/0.31.2 (Revision 775fad2)\t" +
			"reqtime:0.087\t" +
			"taken_sec:0.087\t" +
			"vhost:mackerel.io\t" +
			"apptime:0.021, 0.042", // not a single number
		ContainsString: "ltsv: cannot unmarshal",
	},
}

func TestLTSV_ParseError(t *testing.T) {
	psr := &LTSV{}
	for _, tt := range parseLTSVErrorTests {
		t.Logf("testing: %s", tt.Name)
		if _, err := psr.Parse(tt.Input); err == nil {
			t.Errorf("%s: error should be occured but nil", tt.Name)
		} else if !strings.Contains(err.Error(), tt.ContainsString) {
			t.Errorf("%s: error should be contained %q, but: %s", tt.Name, tt.ContainsString, err)
		}
	}
}

func TestLTSV_ParseError_loose(t *testing.T) {
	psr := &LTSV{Loose: true}
	for _, tt := range parseLTSVErrorTests {
		t.Logf("testing: %s", tt.Name)
		if tt.HardError {
			if _, err := psr.Parse(tt.Input); err == nil {
				if tt.HardError {
					t.Errorf("%s: error should be occured but nil", tt.Name)
				}
			} else if !strings.Contains(err.Error(), tt.ContainsString) {
				t.Errorf("%s: error should be contained %q, but: %s", tt.Name, tt.ContainsString, err)
			}
		} else {
			if _, err := psr.Parse(tt.Input); err != nil {
				if err != nil {
					t.Errorf("%s: error should not be occured but: %s", tt.Name, err.Error())
				}
			}
		}
	}
}


apache_test.go
package axslogparser

import (
	"strings"
	"testing"
)

var benchInputs []string

func init() {
	for _, v := range parseTests {
		if strings.HasPrefix(v.Name, "[Apache]") {
			benchInputs = append(benchInputs, v.Input)
		}
	}
}

func BenchmarkApache_Parse(b *testing.B) {
	p := &Apache{}
	for i := 0; i < b.N; i++ {
		for _, v := range benchInputs {
			p.Parse(v)
		}
	}
}

var parseSuccessTests = []struct {
	Name  string
	Input string
}{
	{
		Name:  "Valid request - many remote host ips",
		Input: `192.168.0.1, 192.168.0.2, 192.168.0.3 10.10.10.10 - - [01/Jan/2005:15:04:05 +0000] "GET /home HTTP/1.1" 200 4589 "https://www.google.com/home" "UA"`,
	},
}

var parseErrorTests = []struct {
	Name           string
	Input          string
	ContainsString string
	HardError      bool
}{
	{
		Name:           "broken",
		Input:          "hoge",
		ContainsString: "(not matched)",
		HardError:      true,
	},
	{
		Name:           "invalid request",
		Input:          `10.0.0.11 - - [11/Jun/2017:05:56:04 +0900] "GET / hoge HTTP/1.1" 200 741 "-" "mackerel-http-checker/0.0.1"`,
		ContainsString: "(invalid request)",
	},
	{
		Name:           "invalid request",
		Input:          `10.0.0.11 - - [11/Jun/2017:05:56:04 +0900] "GET / HTTP/1.1" 200`,
		ContainsString: "(invalid status or size)",
		HardError:      true,
	},
	{
		Name:           "invalid request",
		Input:          `10.0.0.11 - - [11/Jun/2017:05:56:04 +0900] "GET / HTTP/1.1" 2xx 741 "-" "mackerel-http-checker/0.0.1"`,
		ContainsString: "(invalid status:",
	},
}

func TestParse_error(t *testing.T) {
	psr := &Apache{}
	for _, tt := range parseErrorTests {
		t.Logf("testing: %s", tt.Name)
		if _, err := psr.Parse(tt.Input); err == nil {
			t.Errorf("%s: error should be occured but nil", tt.Name)
		} else if !strings.Contains(err.Error(), tt.ContainsString) {
			t.Errorf("%s: error should be contained %q, but: %s", tt.Name, tt.ContainsString, err)
		}
	}
	for _, tt := range parseSuccessTests {
		t.Logf("testing: %s", tt.Name)
		if _, err := psr.Parse(tt.Input); err != nil {
			t.Errorf("%s: error should not be occured but: %s", tt.Name, err.Error())
		}
	}
}

func TestParse_loose(t *testing.T) {
	psr := &Apache{Loose: true}
	for _, tt := range parseErrorTests {
		t.Logf("testing: %s", tt.Name)
		if tt.HardError {
			if _, err := psr.Parse(tt.Input); err == nil {
				if tt.HardError {
					t.Errorf("%s: error should be occured but nil", tt.Name)
				}
			} else if !strings.Contains(err.Error(), tt.ContainsString) {
				t.Errorf("%s: error should be contained %q, but: %s", tt.Name, tt.ContainsString, err)
			}
		} else {
			if _, err := psr.Parse(tt.Input); err != nil {
				if err != nil {
					t.Errorf("%s: error should not be occured but: %s", tt.Name, err.Error())
				}
			}
		}
	}
	for _, tt := range parseSuccessTests {
		t.Logf("testing: %s", tt.Name)
		if _, err := psr.Parse(tt.Input); err != nil {
			t.Errorf("%s: error should not be occured but: %s", tt.Name, err.Error())
		}
	}
}


apache2_test.go
package axslogparser

import (
	"testing"
)

func BenchmarkApache2_Parse(b *testing.B) {
	p := &Apache2{}
	for i := 0; i < b.N; i++ {
		for _, v := range benchInputs {
			p.Parse(v)
		}
	}
}


ltsv.go
package axslogparser

import (
	"strings"
	"time"

	ltsv "github.com/Songmu/go-ltsv"
	"github.com/pkg/errors"
)

// LTSV access log parser
type LTSV struct {
	// If set to true, ignores non-fatal errors while parsing line,
	// which may leave some fields empty or invalid.
	Loose bool
}

// Parse for Parser interface
func (lv *LTSV) Parse(line string) (*Log, error) {
	l := &Log{}
	err := ltsv.Unmarshal([]byte(line), l)
	if err != nil {
		if _, ok := err.(ltsv.UnmarshalError); !(lv.Loose && ok) {
			return nil, errors.Wrapf(err, "failed to parse ltsvlog (not a ltsv): %s", line)
		}
	}
	l.Time, _ = time.Parse(clfTimeLayout, strings.Trim(l.TimeStr, "[]"))
	if err := l.breakdownRequest(); !lv.Loose && err != nil {
		return nil, errors.Wrap(err, "failed to parse ltsvlog (invalid request)")
	}
	return l, nil
}


apache.go
package axslogparser

import (
	"bytes"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/pkg/errors"
)

// Apache log parser
type Apache struct {
	// If set to true, ignores non-fatal errors while parsing line,
	// which may leave some fields empty or invalid.
	Loose bool
}

var logRe = regexp.MustCompile(
	`^(?:(\S+(?:,\s\S+)*)\s)?` + // %v(The canonical ServerName/virtual host) - 192.168.0.1 or 192.168.0.1,192.168.0.2, 192.168.0.3
		`(\S+)\s` + // %h(Remote Hostname) $remote_addr
		`(\S+)\s` + // %l(Remote Logname)
		`([\S\s]+)\s` + // $remote_user
		`\[(\d{2}/\w{3}/\d{2}(?:\d{2}:){3}\d{2} [-+]\d{4})\]\s` + // $time_local
		`(.*)`)

// Parse for Parser interface
func (ap *Apache) Parse(line string) (*Log, error) {
	matches := logRe.FindStringSubmatch(line)
	if len(matches) < 1 {
		return nil, fmt.Errorf("failed to parse apachelog (not matched): %s", line)
	}
	l := &Log{
		VirtualHost:   matches[1],
		Host:          matches[2],
		RemoteLogname: matches[3],
		User:          matches[4],
	}
	if l.Host == "-" && l.VirtualHost != "" {
		l.Host = l.VirtualHost
		l.VirtualHost = ""
		l.User = fmt.Sprintf("%s %s", l.RemoteLogname, l.User)
		l.RemoteLogname = "-"
	}

	l.Time, _ = time.Parse(clfTimeLayout, matches[5])
	var rest string

	l.Request, rest = takeQuoted(matches[6])
	if err := l.breakdownRequest(); !ap.Loose && err != nil {
		return nil, errors.Wrapf(err, "failed to parse apachelog (invalid request): %s", line)
	}
	matches = strings.Fields(rest)
	if len(matches) < 2 {
		return nil, fmt.Errorf("failed to parse apachelog (invalid status or size): %s", line)
	}
	l.Status, _ = strconv.Atoi(matches[0])
	if !ap.Loose {
		if l.Status < 100 || 600 <= l.Status {
			return nil, fmt.Errorf("failed to parse apachelog (invalid status: %s): %s", matches[0], line)
		}
	}
	l.Size, _ = strconv.ParseUint(matches[1], 10, 64)
	l.Referer, rest = takeQuoted(rest)
	l.UserAgent, _ = takeQuoted(rest)
	return l, nil
}

func takeQuoted(line string) (string, string) {
	if line == "" {
		return "", ""
	}
	i := 0
	for ; i < len(line); i++ {
		if line[i] == '"' {
			i++
			break
		}
	}
	if i == len(line) {
		return "", ""
	}
	buf := &bytes.Buffer{}
	escaped := false
	for ; i < len(line); i++ {
		c := line[i]
		if !escaped {
			if c == '"' {
				break
			}
			if c == '\\' {
				escaped = true
				continue
			}
			buf.WriteByte(c)
			continue
		}
		escaped = false
		switch c {
		case 'n':
			buf.WriteByte('\n')
		case 't':
			buf.WriteByte('\t')
		case '\\':
			buf.WriteByte('\\')
		case '"':
			buf.WriteByte('"')
		default:
			buf.WriteByte(c)
		}
	}
	return buf.String(), line[i+1:]
}


apache2.go
package axslogparser

import (
	"bytes"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// Apache2 log parser
type Apache2 struct {
}

var logRe2 = regexp.MustCompile(
	`(?:(?P<vhost>\S+)\s)?` + // %v(The canonical ServerName/virtual host)
		`(?P<remote_addr>\S+)\s` + // %h(Remote Hostname) $remote_addr
		`\S+\s+` + // %l(Remote Logname)
		`(?P<remote_user>\S+)\s` + // $remote_user
		`\[(?P<time_local>\d{2}/\w{3}/\d{2}(?:\d{2}:){3}\d{2} [-+]\d{4})\]\s` + // $time_local
		`(?P<rest>.*)`)

// Parse for Parser interface
func (ap *Apache2) Parse(line string) (*Log, error) {
	matches := logRe2.FindStringSubmatch(line)
	if len(matches) < 1 {
		return nil, fmt.Errorf("not matched")
	}
	var rest string
	l := &Log{}
	for i, name := range logRe2.SubexpNames() {
		switch name {
		case "vhost":
			l.VirtualHost = matches[i]
		case "remote_addr":
			l.Host = matches[i]
		case "remote_user":
			l.User = matches[i]
		case "time_local":
			l.Time, _ = time.Parse(clfTimeLayout, matches[i])
		case "rest":
			rest = matches[i]
		}
	}
	l.Request, rest = takeQuoted(rest)
	matches = strings.Fields(rest)
	if len(matches) > 1 {
		l.Status, _ = strconv.Atoi(matches[0])
		l.Size, _ = strconv.ParseUint(matches[1], 10, 64)
	}
	l.Referer, rest = takeQuoted(rest)
	l.UA, _ = takeQuoted(rest)
	l.breakdownRequest()
	return l, nil
}

func takeQuoted(line string) (string, string) {
	if line == "" {
		return "", ""
	}
	i := 0
	for ; i < len(line); i++ {
		if line[i] == '"' {
			i++
			break
		}
	}
	if i == len(line) {
		return "", ""
	}
	buf := &bytes.Buffer{}
	escaped := false
	for ; i < len(line); i++ {
		c := line[i]
		if !escaped {
			if c == '"' {
				break
			}
			if c == '\\' {
				escaped = true
				continue
			}
		}
		if !escaped {
			buf.WriteByte(c)
			continue
		}
		escaped = false
		switch c {
		case 'n':
			buf.WriteByte('\n')
		case 't':
			buf.WriteByte('\t')
		case '\\':
			buf.WriteByte('\\')
		case '"':
			buf.WriteByte('"')
		default:
			buf.WriteByte(c)
		}
	}
	return buf.String(), line[i+1:]
}


axslogparser.go
package axslogparser

import (
	"fmt"
	"strings"
	"time"
)

// Parser is the interface for accesslog
type Parser interface {
	Parse(string) (*Log, error)
}

// Log is the struct stored parsed result of single line of accesslog
type Log struct {
	VirtualHost     string `ltsv:"vhost"`
	Host            string
	RemoteLogname   string `ltsv:"ident"`
	User            string
	Time            time.Time `ltsv:"-"`
	TimeStr         string    `ltsv:"time"`
	Request         string    `ltsv:"req"`
	Status          int
	Size            uint64
	Referer         string
	UserAgent       string `ltsv:"ua"`
	ReqTime         *float64
	ReqTimeMicroSec *float64 `ltsv:"reqtime_microsec"`
	AppTime         *float64
	TakenSec        *float64 `ltsv:"taken_sec"` // Hatena specific
	ForwardedFor    string
	RequestURI      string `ltsv:"uri"`
	Protocol        string
	Method          string
}

func (l *Log) breakdownRequest() error {
	if l.RequestURI != "" && l.Protocol != "" && l.Method != "" {
		return nil
	}
	stuff := strings.Fields(l.Request)
	if len(stuff) != 3 {
		return fmt.Errorf("invalid request: %s", l.Request)
	}
	if len(stuff) > 0 && l.Method == "" {
		l.Method = stuff[0]
	}
	if len(stuff) > 1 && l.RequestURI == "" {
		l.RequestURI = stuff[1]
	}
	if len(stuff) > 2 && l.Protocol == "" {
		l.Protocol = stuff[2]
	}
	return nil
}

const clfTimeLayout = "02/Jan/2006:15:04:05 -0700"

// Parsers represents a set of available parsers
type Parsers struct {
	Apache *Apache
	LTSV   *LTSV
}

// GuessParser guesses the parser from line
func (ps Parsers) GuessParser(line string) (Parser, *Log, error) {
	if strings.Contains(line, "\thost:") || strings.Contains(line, "\ttime:") {
		l, err := ps.LTSV.Parse(line)
		if err == nil {
			return ps.LTSV, l, err
		}
	}
	l, err := ps.Apache.Parse(line)
	if err != nil {
		return nil, nil, err
	}
	return ps.Apache, l, nil
}

// Parse log line
func (ps Parsers) Parse(line string) (*Log, error) {
	_, l, err := ps.GuessParser(line)
	return l, err
}

// GuessParser guesses the parser from line, uses default parsers for each format
func GuessParser(line string) (Parser, *Log, error) {
	ps := Parsers{
		LTSV:   &LTSV{},
		Apache: &Apache{},
	}
	return ps.GuessParser(line)
}

// Parse log line using default parsers
func Parse(line string) (*Log, error) {
	_, l, err := GuessParser(line)
	return l, err
}
